[Beebeeoii] iP

Beebeeoii opened the pull request:
# Membot

> "The only to-do list manager that you will ever need" ~ someone

Membot is a friendly to-do list manager that supports CLI for those keyboarders out there !

## Features

- Add to-do items that includes a _deadline_, _starting dateTime_ or _ending date/time_
- Save to-do list to storage
- Load to-do list from storage
- ~~Fast~~ **_Super_ fast** to use

## Installation

1. Download the `.jar` executable from [packages](https://github.com/Beebeeoii/ip/releases).
2. Run the executable.
3. Start managing your tasks with ease 😉

## Future Plans

- [x] Implement a GUI

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
feat: created a print wrapper class

Commit:
feat: added greet, echo, exit functionality

Commit:
chore: renamed duke to membot

Commit:
fix: loop did not break with terminate key

Commit:
feat: added Task model

Commit:
feat: added ability to create and list tasks

Commit:
ui: tidied up UI for listing of tasks

Commit:
code: closed scanner upon main exit

Commit:
feat: input validator for check/unchecking tasks

Commit:
feat: added done/undone status for tasks

Commit:
refactor: replaced factory method for Tasks with constructor

Commit:
feat: added deadline, event, todo

Commit:
ui: improved ui when list of tasks is empty

Commit:
fix: user input handling for deadline and event

Commit:
test: updated test

Beebeeoii committed:
feat: added custom exceptions

Beebeeoii committed:
feat: added error printing

Beebeeoii committed:
feat: added error handling for inputs

Beebeeoii committed:
feat: added deletion of tasks

Beebeeoii committed:
test: updated tests

Commit:
feat: add saving tasks to file

Commit:
feat: add loading of tasks from file

Commit:
feat: add recognising date object from string

Commit:
feat: add recognising date object from string

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

Commit:
refactor: add more abstraction for commands

Commit:
test: add test for listPrint

Beebeeoii committed:
refactor: replace IndexOutOfBoundsEx with InvalidCommandEx

Beebeeoii committed:
fix: fix indexing for deadline command input

Beebeeoii committed:
fix: fix indexing for deadline command input

Beebeeoii committed:
refactor: remove unnecessary IndexOutOfBoundsEx

Beebeeoii committed:
refactor: privatise methods used only by self

Beebeeoii committed:
test: add test cases for Printer

Beebeeoii committed:
test: add test cases for utils

Beebeeoii committed:
chore: update gradle config

Beebeeoii committed:
docs: add documentations

Beebeeoii committed:
style: fix style

Beebeeoii committed:
feat: add ability to find task using keyword

Beebeeoii committed:
Merge branch 'branch-A-CodingStandard'

Beebeeoii committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/commands/Command.java
#	src/main/java/model/Task.java

Commit:
style: fix style

Commit:
chore: updated dependencies to include JavaFX

zenithyap commented on src/test/java/utils/InputValidatorTest.java:
```diff
@@ -0,0 +1,244 @@
+        InvalidCommandException e3 = assertThrows(InvalidCommandException.class, () ->
+                InputValidator.normaliseDeadlineInput(input3));
+        assertEquals("[Invalid command] Invalid Syntax - "
+                        + "\"deadline [title] /by [deadline]\" (e.g. \"deadline physics project /by tomorrow 3pm\"",
```
You could separate this line into two as it is quite long, similar to line 116.

zenithyap commented on src/main/java/model/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+ * to be completed by.
+ */
+public class Deadline extends Task {
+    public static final String TAG = "[D]";
```
I like the use of a variable to name the tags.

zenithyap commented on src/main/java/utils/InputValidator.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * Cleanses and normalises an <code>event</code> command.
```
Description could be more descriptive.

zenithyap commented on src/main/java/utils/InputValidator.java:
```diff
@@ -0,0 +1,220 @@
+        return res;
+    }
+
+    private static String extractDeadline(String input) throws NoDeadlineFoundException {
```
Please add javadocs for this and subsequent methods.

zenithyap commented on src/test/java/utils/InputValidatorTest.java:
```diff
@@ -0,0 +1,244 @@
+package utils;
```
I feel that variable names could be more descriptive for the inputs and exceptions.

zenithyap commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,46 @@
+list
```
Good job on the comprehensive test cases.

Commit:
ui: implement gui

Commit:
wip: implement gui

Commit:
ui: implement gui

alson001 commented on src/main/java/model/Task.java:
```diff
@@ -0,0 +1,249 @@
+     * @param in The array of data to be loaded.
+     */
+    public static void load(ArrayList<String> in) {
+        Base64.Decoder d = Base64.getDecoder();
```
Maybe name it decoder? d is quite ambiguous

alson001 commented on src/main/java/model/Task.java:
```diff
@@ -0,0 +1,249 @@
+     * @throws IOException If some sort of IO error occurs during the process of writing.
+     */
+    public static void save(Outputable out) throws IOException {
+        Base64.Encoder e = Base64.getEncoder();
```
maybe name it encoder?

alson001 commented on src/main/java/Membot.java:
```diff
@@ -0,0 +1,78 @@
+        } catch (IOException e) {
+            ui.printlnError(e.toString());
+        }
+        ui.println("Welcome to\n" + LOGO);
```
maybe can move "Welcome to\n" to UI too

alson001 commented on src/main/java/Membot.java:
```diff
@@ -0,0 +1,78 @@
+            ui.printlnError(e.toString());
+        }
+        ui.println("Welcome to\n" + LOGO);
+        ui.println("How may I assist you today?");
```
Maybe can move this statement to UI under ui.greetings()

alson001 reviewed (commented):
Overall this bot is very unique. I think you put in a lot of effort as this is really different from a standard bot in how the classes interact with each other. The test cases are plenty too. Well done. LGTM

Commit:
ui: implement gui

Commit:
ui: remove CLI printer

Commit:
ui: clean up printing of messages

Commit:
style: clean up

Commit:
feat: implement HELP command

Commit:
feat: implement HELP command

Commit:
ui: reduce EXIT_DELAY to 500

Commit:
ui: reduce EXIT_DELAY to 500

Commit:
style: rename encoder and decoder for clarity

Commit:
style: utilise assertions

Assertions: utilise `assert` for empty `default` clauses in `switch-case`

`default` clauses throw an error or return arbitruary values.

Using `assert` is more readable and ensures that errors are caught in the future when the `switch-case` statements are not updated.

Let's:
- ensure that `default` clauses are updated when new values are added for the `switch-case` statements in the future

Commit:
refactor: remove `isFindInputValid` method

Refactor: remove redundant `isFindInputValid` method from `InputValidator.java`.

`isFindInputValid()` is used to validate whether the `find` command syntax is valid.

There is already a `isCheckInputValid()` method which does essentially the same thing, just with an extra check for the `int`.

It is better to maintain just one of such validator method as it will maintain a single level of abstraction. Changing such a method in the future would just require us to change one method, ensuring that there is fewer points of errors.

Let's:
- remove `isFindInputValid()` and rename `isCheckInputValid()` to a more general name: `isSingleInputValid()`

Commit:
chore: update README

Commit:
Merge branch 'branch-Level-10'

Commit:
chore: update README

Commit:
Merge pull request #3 from Beebeeoii/branch-A-Assertions

refactor: utilise assertions

Commit:
Merge pull request #5 from Beebeeoii/branch-A-CodeQuality

refactor: remove `isFindInputValid` method

Commit:
chore: gradle workflow

Commit:
refactor: update `isSingleInputValid` to take in input with spaces

Refactor: include an additional parameter `isMulti` for the method `isSingleInputValid` to indicate if the input consists of multiple words delimited by `space` characters.

Method was only able to take inputs of one word which causes `find` commands to be inaccurately judged to be invalid.

By including the extra parameter, it can be used for all the commands that take in a single input, such as `done`, `undone`, `find`, `delete`.

Let's:
- utilise `isSingleInputValid` for all commands that take in a single input, even if the input might be a sentence made up of words delimited by `space` characters

Commit:
test: update test for `isSingleInputValid`

Commit:
chore: update `mainClassName` to `membot.Launcher`

Commit:
ci: update with gradle checkstyle and test

Commit:
Merge pull request #6 from Beebeeoii/branch-A-CI

implement CI to Membot

Commit:
refactor: utilise `Stream` for `find()` and `listAll()`

Commit:
feat: add `Sort` command for date, status, title, type

Commit:
style: fix `switch` statement style

Commit:
Merge pull request #7 from Beebeeoii/branch-C-Sort

feat: add `Sort` command for date, status, title, type

Commit:
chore: update `archiveVersion`

Commit:
test: add tests for `Task` related

Commit:
test: add tests for `UiPrinter`

Commit:
fix: change error thrown by `deleteLast()`

Commit:
fix: `Task` list not being emptied before tests

Commit:
style: removed unused import

Commit:
style: removed unused variable

Commit:
test: add `Command` tests

Commit:
test: add `tearDown()` after creating tasks

Commit:
refactor: change constructor to be `protected`

Commit:
asset: add new icon for Membot

Commit:
asset: add new icon for Membot

Commit:
refactor: extract magic string and update names

Commit:
refactor: extract magic string and update names

Commit:
fix: set min width and height for window

Commit:
fix: set min width and height for window

Commit:
ui: add icon for membot

Commit:
ui: add icon for membot

Commit:
ui: add minor changes to ui

Commit:
ui: add minor changes to ui

Commit:
refactor: remove unnecessary comment

Commit:
refactor: remove unnecessary comment

Commit:
ui: add colour for error messages

Commit:
ui: add colour for error messages

Commit:
Merge pull request #10 from Beebeeoii/branch-A-BetterGui

ui: improve general UI

Commit:
feat: add ability to create `Membot` instance for test env

Commit:
test: update test to incorporate `Message`

Commit:
Merge pull request #9 from Beebeeoii/branch-A-MoreTesting

test: add more tests

Commit:
refactor: remove hardcoding of sort options

Commit:
ui: improve UI for `help`, `todo`, `deadline`, `event`

Commit:
ui: update checkmark for done tasks

Commit:
docs: add `Membot` screenshot

Commit:
refactor: change `bye` to `exit`

Commit:
docs: update README

Commit:
docs: update image max width

Commit:
docs: update image max width

Commit:
docs: update image center align

Commit:
docs: update README.md

Commit:
docs: update README.md

Commit:
docs: update README.md

Commit:
feat: tasks are automatically after every command

Commit:
chore: bump version to v0.3

Commit:
refactor: improve code quality for `sort` execute
